File: native/android/app/src/main/java/com/audiovisualizer/UnityBridge.kt
package com.audiovisualizer

import android.content.Context
import com.unity3d.player.UnityPlayer
import org.json.JSONObject

object UnityBridge {
    private var unityPlayer: UnityPlayer? = null
    private var isInitialized = false

    fun initialize(context: Context) {
        if (isInitialized) return
        isInitialized = true
    }

    fun setUnityPlayer(player: UnityPlayer) {
        unityPlayer = player
    }

    fun sendMessage(gameObject: String, method: String, message: String) {
        UnityPlayer.UnitySendMessage(gameObject, method, message)
    }

    fun sendTrackMessage(
        source: String,
        title: String,
        artist: String,
        artworkUrl: String? = null,
        durationMs: Int
    ) {
        val json = JSONObject().apply {
            put("type", "track")
            put("source", source)
            put("title", title)
            put("artist", artist)
            put("durationMs", durationMs)
            artworkUrl?.let { put("artworkUrl", it) }
        }
        
        sendMessage("NativeBridge", "OnNativeMessage", json.toString())
    }

    fun sendPlaybackMessage(
        isPlaying: Boolean,
        positionMs: Int,
        volume: Float? = null
    ) {
        val json = JSONObject().apply {
            put("type", "playback")
            put("isPlaying", isPlaying)
            put("positionMs", positionMs)
            volume?.let { put("volume", it) }
        }
        
        sendMessage("NativeBridge", "OnNativeMessage", json.toString())
    }

    fun sendBandsMessage(
        bass: Float,
        mid: Float,
        high: Float,
        sub: Float? = null,
        kick: Float? = null,
        energy: Float? = null
    ) {
        val json = JSONObject().apply {
            put("type", "bands")
            put("bass", bass.toDouble())
            put("mid", mid.toDouble())
            put("high", high.toDouble())
            sub?.let { put("sub", it.toDouble()) }
            kick?.let { put("kick", it.toDouble()) }
            energy?.let { put("energy", it.toDouble()) }
        }
        
        sendMessage("NativeBridge", "OnNativeMessage", json.toString())
    }

    fun sendPresetMessage(
        presetName: String,
        intensity: Float? = null,
        speed: Float? = null,
        trailsOn: Boolean? = null,
        trailsAmount: Float? = null
    ) {
        val json = JSONObject().apply {
            put("type", "preset")
            put("presetName", presetName)
            intensity?.let { put("intensity", it.toDouble()) }
            speed?.let { put("speed", it.toDouble()) }
            trailsOn?.let { put("trailsOn", it) }
            trailsAmount?.let { put("trailsAmount", it.toDouble()) }
        }
        
        sendMessage("NativeBridge", "OnNativeMessage", json.toString())
    }

    fun sendControlMessage(action: String, value: Float? = null) {
        val json = JSONObject().apply {
            put("type", "control")
            put("action", action)
            value?.let { put("value", it.toDouble()) }
        }
        
        sendMessage("NativeBridge", "OnNativeMessage", json.toString())
    }
}

File: native/ios/AudioVisualizer/UnityBridge.swift
import Foundation
import UnityFramework

@objc public class UnityBridge: NSObject {
    private static var instance: UnityBridge?
    private var unityFramework: UnityFramework?
    private var isInitialized = false
    
    public static var shared: UnityBridge {
        if instance == nil {
            instance = UnityBridge()
        }
        return instance!
    }
    
    private override init() {
        super.init()
    }
    
    public func initialize() {
        guard !isInitialized else { return }
        
        let bundlePath = Bundle.main.bundlePath
        let frameworkPath = bundlePath + "/Frameworks/UnityFramework.framework"
        
        guard let bundle = Bundle(path: frameworkPath) else {
            print("[UnityBridge] Failed to load Unity framework bundle")
            return
        }
        
        if !bundle.isLoaded {
            bundle.load()
        }
        
        guard let principalClass = bundle.principalClass as? UnityFramework.Type else {
            print("[UnityBridge] Failed to get UnityFramework class")
            return
        }
        
        unityFramework = principalClass.getInstance()
        
        if let framework = unityFramework {
            framework.setDataBundleId("com.unity3d.framework")
            framework.register(self)
            framework.runEmbedded(withArgc: CommandLine.argc, argv: CommandLine.unsafeArgv, appLaunchOpts: nil)
            isInitialized = true
        }
    }
    
    public func showUnity() {
        unityFramework?.showUnityWindow()
    }
    
    public func sendMessage(_ gameObject: String, method: String, message: String) {
        unityFramework?.sendMessageToGO(withName: gameObject, functionName: method, message: message)
    }
    
    public func sendTrackMessage(source: String, title: String, artist: String, artworkUrl: String?, durationMs: Int) {
        var dict: [String: Any] = [
            "type": "track",
            "source": source,
            "title": title,
            "artist": artist,
            "durationMs": durationMs
        ]
        
        if let artwork = artworkUrl {
            dict["artworkUrl"] = artwork
        }
        
        if let json = try? JSONSerialization.data(withJSONObject: dict),
           let jsonString = String(data: json, encoding: .utf8) {
            sendMessage("NativeBridge", method: "OnNativeMessage", message: jsonString)
        }
    }
    
    public func sendPlaybackMessage(isPlaying: Bool, positionMs: Int, volume: Float? = nil) {
        var dict: [String: Any] = [
            "type": "playback",
            "isPlaying": isPlaying,
            "positionMs": positionMs
        ]
        
        if let vol = volume {
            dict["volume"] = vol
        }
        
        if let json = try? JSONSerialization.data(withJSONObject: dict),
           let jsonString = String(data: json, encoding: .utf8) {
            sendMessage("NativeBridge", method: "OnNativeMessage", message: jsonString)
        }
    }
    
    public func sendBandsMessage(bass: Float, mid: Float, high: Float, sub: Float? = nil, kick: Float? = nil, energy: Float? = nil) {
        var dict: [String: Any] = [
            "type": "bands",
            "bass": bass,
            "mid": mid,
            "high": high
        ]
        
        if let s = sub { dict["sub"] = s }
        if let k = kick { dict["kick"] = k }
        if let e = energy { dict["energy"] = e }
        
        if let json = try? JSONSerialization.data(withJSONObject: dict),
           let jsonString = String(data: json, encoding: .utf8) {
            sendMessage("NativeBridge", method: "OnNativeMessage", message: jsonString)
        }
    }
    
    public func sendPresetMessage(presetName: String, intensity: Float? = nil, speed: Float? = nil, trailsOn: Bool? = nil, trailsAmount: Float? = nil) {
        var dict: [String: Any] = [
            "type": "preset",
            "presetName": presetName
        ]
        
        if let i = intensity { dict["intensity"] = i }
        if let s = speed { dict["speed"] = s }
        if let t = trailsOn { dict["trailsOn"] = t }
        if let ta = trailsAmount { dict["trailsAmount"] = ta }
        
        if let json = try? JSONSerialization.data(withJSONObject: dict),
           let jsonString = String(data: json, encoding: .utf8) {
            sendMessage("NativeBridge", method: "OnNativeMessage", message: jsonString)
        }
    }
}

extension UnityBridge: UnityFrameworkListener {
    public func unityDidUnload(_ notification: Notification!) {
        isInitialized = false
        unityFramework = nil
    }
}
